### ISSUE TITLE
Neovide Windows freezes when quitting

### ISSUE THREAD

**Issue by Parsifa1** [neovide/neovide#2182]
## **Describe the bug**

Neovide Windows 11 WSL2 not responding when quit

## **To Reproduce**

1. Open WSL terminal
1. Run ` neovide.exe --wsl `
1. Quit neovide either using ` :q ` or clicking close button
1. Neovide becomes freezed

## **Expected behavior**

exit neovim and neovide as expected

## **Screenshots**

![image](https://github.com/neovide/neovide/assets/31800073/95afc09d-c22b-4d2d-ba75-d0e3491e904e)

## **Desktop (please complete the following information):**

 - OS: Windows11 23H2
 - Neovide Version : 0.11.2
 - Neovim Version: 
 ` nightly `
 ```
 NVIM v0.10.0-dev-1846+gf31f260f0c6
Build type: RelWithDebInfo
LuaJIT 2.1.1702233742
Run "nvim -V1 -v" for more info
```

 **Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/13687097/neovide_rCURRENT.log)

**Comment by fredizzimo** [comment]
I'm using Neovide  in Windows and WSL all the time, and don't see the problem, so it's hard to tell what's wrong.

One possibility is the clipboard hang, described here https://github.com/neovide/neovide/issues/2003. On WSL you only need Win32yank.exe, so if you have xlclip or similar installed, perhaps you could try to uninstall that. It also causing problems with the copy and paste functionality, since it has a higher priority than Win32Yank, or at least it used to have when I last checked.

**Comment by Parsifa1** [comment]
In fact, I only used win32yank, and if I didn't use win32yank, ^M would appear on each line due to the difference between Windows and Linux.

![image](https://github.com/neovide/neovide/assets/31800073/51d0aa39-0f7a-4628-a95d-900cc04dab48)

**Comment by Parsifa1** [comment]
I also tried close ` vim.opt.clipboard = "unnamedplus" `, and it comes again

![image](https://github.com/neovide/neovide/assets/31800073/cbc23ecc-069e-4455-9fae-18a9bd3ff608)

**Comment by Parsifa1** [comment]
> @Parsifa1, can you confirm that it's clipboard related by not setting `vim.opt.clipboard = "unnamedplus"` at all? If that works, then we can close this one and track everything under the same issue，你能通过根本不设置 `vim.opt.clipboard = "unnamedplus"` 来确认它与剪贴板相关吗？如果可行，那么我们可以关闭此问题并跟踪同一问题下的所有内容
> 
> * [Neovide doesn't close until after clipboard has been changed externally #2003Neovide 直到剪贴板被外部更改后才会关闭 #2003](https://github.com/neovide/neovide/issues/2003)

it seems that when i close all of the ` vim.opt.clipboard = "unnamedplus" `,everything runs well!

thank you so much!!

**Comment by Parsifa1** [comment]
Oh, it seems that things are not that simple. After I close all of the `vim.opt.clipboard = "unnamedplus" `, I still cannot use ` :q` to exit.

![image](https://github.com/neovide/neovide/assets/31800073/f02229c1-8f93-4208-ba68-b446e38272b1)

**Comment by rytc** [comment]
I'm seeing this issue on windows 10 without WSL on the latest 0.12 release

**Comment by fredizzimo** [comment]
Unfortunately, I'm unable to repeat this on any of my Windows 11 machines. Neither with `--wsl` or without. 

So, it would be good if someone, who can, could check with either a debugger or with log statements to determine exactly how far it goes in the progress, and what it's waiting for.

**Comment by comiluv** [comment]
I've been getting freezes since lastest update as well.
Windows 11 without wsl, specifically version 10.0.22631

**Comment by comiluv** [comment]
I don't have neovide log files generated at the moment of crashing but I have some report.wer files generated by Windows when neovide crashed, which were found in `C:\ProgramData\Microsoft\Windows\WER\ReportArchive` I can post the files if it could help reveal the problem

**Comment by fredizzimo** [comment]
It might help, but I think a callstack with all the threads visible from a debug build will be needed. In VSCode, you can use the following launch configuration.

```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch",
            "type": "lldb",
            "request": "launch",
            "args": [],
            "cargo": {
                "args": ["build"]
            }
        },
    ]
```

Together with https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb, if you otherwise know how to use the VSCode debugger.

A nvim dap configuration was also recently added https://github.com/neovide/neovide/blob/main/.nvim-dap.lua, but I have not tried debugging using that myself.

**Comment by Oreolek** [comment]
I also see this but on Linux (not sure if that's the same issue or a new one).

**Comment by fredizzimo** [comment]
I don't have high hopes that this fixes it, but it's possible, so could you try that? I would like more testing for it anyway.
* https://github.com/neovide/neovide/pull/2215,

**Comment by yngwi** [comment]
I have the same problem on Windows 11.

**Comment by amiorin** [comment]
I have the same problem with macOS Sonoma 14.2.1.
https://youtu.be/px_f1t69Zhg

**Comment by sid-6581** [comment]
I have the same problem all the time again, like I did before. From a quick check in the debugger it appears to be the same issue as #2130. neovide is stuck waiting for the `NeovimRuntime` drop.

**Comment by gilbh** [comment]
I have the same issue with 
Windows 11 Pro 
Version 10.0.22621 Build 22621

**Comment by fredizzimo** [comment]
Would it be possible for someone that can reproduce this to compile to compile tokio from source and see if this is triggered

https://github.com/tokio-rs/tokio/blob/84c5674c601dfc36ab417ff0ec01763c2dd30a5c/tokio/src/process/windows.rs#L166-L169

That's the code that detects if the process has exited. We need to find out exactly what Neovide is waiting for, and based on @sid-6581's previous investigation, that should be a good place to start.

**Comment by fredizzimo** [comment]
This was also posted here https://github.com/neovide/neovide/pull/2130#issuecomment-1813525592

![image](https://github.com/neovide/neovide/assets/28602336/7ec29aa4-d21a-4fee-b18c-a0569fc3b3a9)

It would be useful to know what the last line expands to, but it looks like it's stuck reading from the input stream.

**Comment by sid-6581** [comment]
I'm seeing a lot of new freezes when using neovide, the same quitting freeze mentioned here, but also while using it. I suspect they're separate issues, but I'll try to debug everything I'm seeing.

**Comment by fredizzimo** [comment]
It's strange how different it can be on different systems, I don't remember when the last freeze I saw on Windows 11, exit or otherwise, and I'm using it full-time at work.

But on the other hand, I have't really used the exact release 0.12.0 much, and 0.12.1 not at all. So there might be something in the very latest PRs we merged.

**Comment by sid-6581** [comment]
I'm seeing the freezes consistently on multiple systems, desktops and laptops. At least I can reproduce the issues, but the freezing when quitting is a tricky one to figure out the actual cause of.

**Comment by sid-6581** [comment]
Not waiting forever for the tokio runtime will definitely fix it, because once the wait times out, neovide will for sure exit. That's the only thing holding it up, waiting for the neovim task pointlessly after the process is gone.

But, it is an annoying workaround and I hope to find a cleaner way. I'll be playing around with it this week and see if I can get somewhere.

**Comment by fredizzimo** [comment]
But before that commit we called std::process::exit without any waiting and you said this:
> Before https://github.com/neovide/neovide/commit/2a1074b3610c07b637e6bce8afb48fccf0767e7b the window would be closed when neovim quits, but if the stdin/stdout pipes aren't broken even when the neovim process is gone, the neovide process just hangs around in the background and never exits

So it seems like even a forced exit isn't enough, or did I misunderstand the comment?

**Comment by sid-6581** [comment]
The commit I referenced was after #2035, so it was waiting on the runtime at that point. #2035 was merged on Oct 1, but the commit I mentioned was 3 weeks ago.

**Comment by fredizzimo** [comment]
Ah sorry. I looked at the wrong commit with too many tabs open on my ipad.

Yes, that commit closed the window earlier, due to the render thread, which owned it, was joined.

**Comment by sid-6581** [comment]
No worries! I'll get to the bottom of it one way or the other. Out of necessity if nothing else, lol.

**Comment by fredizzimo** [comment]
The first part is something we want to do anyway. At last on macOS, the window should be created inside the winit event loop and also not outlive it. More specifically it should be created after `Event::NewEvents(StartCause::Init)`, but it's not very well documented
https://github.com/rust-windowing/winit/issues/2051#issuecomment-961079120, https://github.com/rust-windowing/winit#macos

But in the case of Neovide, we actually want to do it inside this 
https://github.com/neovide/neovide/blob/bbb26e84ac0d76a44ec71c53f7e104bc37054b9a/src/window/window_wrapper.rs#L405. So instead of showing the window, we would also create it there.

I'm planning to do that in maybe two weeks, but if anyone wants to do it before that, feel free to do so.

**Comment by fredizzimo** [comment]
I have been trying to repeat this in many different ways with both wsl and without, but unsuccessfully. It does not crash even when I add additional rpc calls when Neovim  is about to exit. And I'm using the latest master without the proposed fix https://github.com/neovide/neovide/pull/2265, 

I wonder if it's something specific to the configuration or plugins used. **Does it also hang if you run `neovide -- --clean`?**

I did notice another problem though, sometimes especially when I quit with `:q` the quit request from Neovim is not received, the process exits before the message is sent to Neovide. This means that the exit code will be wrong when quitting with `:cquit`. But we are doing it the recommended way https://github.com/neovim/neovim/issues/7727, I don't see any way to guarantee that the notification is actually received before neovim is allowed to close. The proposed fix https://github.com/neovide/neovide/pull/2265, probably makes this even worse by not waiting at all.

In that case the log displays just
> INFO [neovide::running_tracker] Quit neovim processed failed

When I close the window, the request is usually reeceived and the output is
> TRACE [neovide::bridge::handler] Neovim notification: "neovide.quit"
INFO [neovide::running_tracker] Quit with code 0: Quit from neovim
INFO [neovide::running_tracker] Quit neovim processed failed

NOTE: In the log from the first message in this issue, the output is the following
> TRACE [neovide::bridge::handler] Neovim notification: "neovide.quit"
INFO [neovide::running_tracker] Quit with code 0: Quit from neovim

So, the `Quit neovim processed failed` message is missing, which makes sense, since that output when Neovide detects that the Neovim process is gone.

**Comment by fredizzimo** [comment]
Adding `vim.cmd("sleep 1")` after https://github.com/neovide/neovide/blob/3e575fcd21863fdaa1fd14d23db649738694af97/lua/init.lua#L97

Fixes the exit code issue, but does not cause any hangs for me either.

A shorter sleep is not consistent enough, but changing the code to do a rpc request instead of a notify is a better fix, but still no hangs.

**Comment by yngwi** [comment]
Hi @fredizzimo. For me Neovide crashes very often on exit but I couldn’t reproduce any crashes with `neovide -- --clean` so far.  So your guess that it might be the config could be correct. I don’t know what it could be, though. Any idea how to debug? I tried to successively remove all plugins and add them one after each one but I couldn‘t reproduce it exactly, it apparently crashes with different combinations of plugins. What I do seem to have noticed is the following:

I use [lazy.nvim](https://github.com/folke/lazy.nvim) to manage my Neovim plugins with many of them set to be lazily loaded. I have tried removing all plugins that are loaded for one reason or another on start and with no such plugins I can’t seem to reproduce the crashes, so it might have to do something with the things lazy.nvim does? Does that make any sense? Of course it can still be some combination of plugins. I‘ll try to continue investigating.

**Comment by fredizzimo** [comment]
If you are able to share your lazy config, I might be able to reproduce it, but in the end I think it's some subtle timing issues that triggers when the stars align correctly.

It's possible that I found the root cause earlier today when I debugged another issue, there appears to be a bug in the Rust standard library for windows https://github.com/neovide/neovide/issues/2299#issuecomment-1907411309. At least that bug can explain this behaviour if it causes `WaitForSingleObject` to block forever.

**Comment by yngwi** [comment]
Here is my config: https://gitlab.nettek.at/djeller/neovim-dotfiles, lazy is here: https://gitlab.nettek.at/djeller/neovim-dotfiles/-/blob/main/lua/lazynvim.lua
The only specialty that I’m aware is that I move the install location.

**Comment by fredizzimo** [comment]
Thanks, I tried the config but did not see any hangs on my system, so I guess we have to wait and see if someone else can repeat it and debug properly or if the standard library bug is the actual issue.

Meanwhile we do have this workaround:
* https://github.com/neovide/neovide/pull/2265

**Comment by fredizzimo** [comment]
* https://github.com/neovide/neovide/pull/2265

Has been merged, but I don't consider the issue fixed. The biggest blocker for making a proper fix, is that I can't repeat the issue myself, I have tried this on 3 different Windows 11 machines, and none of them hangs. Are there any special steps or anything that could explain why it hangs for some people? NOTE: For testing the issue you can no longer use Neovide main, use this commit instead 936dfe25b61c03cfcbd0be9cfebc35187232caa9

I'm quite sure the actual issue is in the rust standard library and the use of WaitForSingleObject, waiting for a file handle instead of an EventObject. See this message https://github.com/neovide/neovide/issues/2299#issuecomment-1907411309. 

I can make a fix for that, but in order for someone to test it out, you need to be able to compile and use the rust toolchain from source, see:
https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html
and
https://github.com/rust-lang/rust/blob/master/INSTALL.md

**Comment by tusan-tsvetkoff** [comment]
I am having the absolute same issue, unfortunately, but I also suspect it stems from the fact that when running normal neovim from Windows Terminal (Powershell) and quitting, it "freezes" until I press a button, once I do that, it goes back to the terminal. 

With Neovide, running it through the terminal `neovide` and then exiting with `:q` or `:wqa`, it straight up freezes as mentioned above.

Nvim version:
NVIM v0.10.0-dev-0a3505e
Build type: RelWithDebInfo
LuaJIT 2.1.1696795921

Windows 10
Latest Powershell version.

[My nvim config.](https://github.com/tusan-tsvetkoff/neovim_config)

**Comment by rytc** [comment]
I decided to try to do a bit more trouble shooting on this. I started by running Neovide with `--clean`. Quitting while running with `--clean` didn't freeze. I then deleted my plugins from the AppData directory and then comment out the lines in my `init.vim` to see if I could narrow down a config option that might be causing this issue. Then went through and uncommented the config options, and then uncommented and reinstalled my plugins one by one.

Surprisingly, doing this stopped the crashes! I'm suspecting there is something cached from the previous version of neovide that wasn't migrated over correctly? Maybe reinstalling the plugins through neovide changed something? I'm not familiar with the internals so I can't really say, but doing this fixed it for me.

**Comment by fredizzimo** [comment]
It does make sense that a corrupted shada would cause Neovide to freeze and I guess what @rytc did could have caused the shada to fix itself again.

But it does not quite explain the behaviour that @sid-6581 was reporting, where it looked like Neovim had actually quit, and only the io pipe was open. Although I guess it could cause some very late message to get sent to Neovide and that actually causes the hang. I will try to corrupt the shada data and see what happens.

**Comment by sid-6581** [comment]
> Hi @fredizzimo. For me Neovide crashes very often on exit but I couldn’t reproduce any crashes with `neovide -- --clean` so far. So your guess that it might be the config could be correct. I don’t know what it could be, though. Any idea how to debug? I tried to successively remove all plugins and add them one after each one but I couldn‘t reproduce it exactly, it apparently crashes with different combinations of plugins. What I do seem to have noticed is the following:
> 
> I use [lazy.nvim](https://github.com/folke/lazy.nvim) to manage my Neovim plugins with many of them set to be lazily loaded. I have tried removing all plugins that are loaded for one reason or another on start and with no such plugins I can’t seem to reproduce the crashes, so it might have to do something with the things lazy.nvim does? Does that make any sense? Of course it can still be some combination of plugins. I‘ll try to continue investigating.
> 
> Edit: scratch that, I just got a crash with only lazy and no plugin being loaded.

When you say crash, do you actually mean freeze? If it freezes, is the neovim process still running? If it's a crash it sounds like a separate problem that should have a new issue opened.

**Comment by fredizzimo** [comment]
If someone with the hang could revert 28cf508d562840bbd1bd329fe1b4bacdfa92ff54, abd test with https://github.com/neovide/neovide/pull/2336, it would be good.

I have a feeling that it fixes the root cause. Maybe not for the corrupted shada files, but the other hangs.

**Comment by Parsifa1** [comment]
I tried the latest https://github.com/neovide/neovide/commit/71e9fa49fcb393662ee03e42968724b91cde8e62, which seems to work smoothly. I will close this issue. Feel free to reopen it if you need.

**Comment by fredizzimo** [comment]
I'm re-opening this, since @sid-6581, has reported that it still hangs with Neovide 0.13.1 here
* https://github.com/neovide/neovide/pull/2463

**Comment by fredizzimo** [comment]
So far, I only know that @sid-6581 still has this problem, but if other people has the prolem, please let me know. It would also be good to know the exact Windows version number, in case this is a problem to some specific version of Windows. See also https://github.com/neovide/neovide/issues/2299, which I think is related.

**Comment by yngwi** [comment]
I have the problem as well. I didn’t have much time due to work but for me it seemed as if certain processes started with nvim like node were still running after nvim/neovide was closed (either with :q or the close button). When killing those in the task manager, the neovide window would close as well.

**Comment by fredizzimo** [comment]
@yngwi

That seems like a possible explanation. Neovim itself should wait for the subprocesses from exiting, and if they don't neither Neovim or Neovide will exit completely. It's strange that it does not happen in terminal nvim, but maybe there's a similar issue to the one we had with the clipboard on Linux and Mac
* https://github.com/neovim/neovim/pull/28157

In that case the stdio and stderr were inherited by the child process, which kept them open and blocked the exit detection.

I don't think the exact same case ever applies to Windows, but something similar might happen.

**Comment by DeadRobotDev** [comment]
It's impacting me as well and I've also noted similar behaviour to @yngwi , specifically with BiomeJS.

If I run Neovide and close before opening a buffer, it closes fine.
If I open a buffer and the Biome LSP starts, it will freeze on close. Forcefully closing the Biome process in Task Manager fixes this.

If I disable the Biome LSP in my config so it's just TypeScript, it closes fine.
If I just open other filetypes and start their LSPs, such as Markdown and Go, it closes fine.

Windows: 23H2 22631.3880
Neovim: v0.11.0-dev-578+gb04b263e1 (I've also experienced it with 0.10.0)
Neovide: 0.13.3
BiomeJS: 1.8.2

**Comment by yngwi** [comment]
Yes, it seems as if some tools that are used in language servers or similar things (for me it’s node, among others) are not being closed when nvim is run in Neovide. Unfortunately, I don’t know what I could do to help fix this. When I work on a rust project, I also have to force quit the leftover things from the task manager most of the time, even without node.

**Comment by fredizzimo** [comment]
Are you running Neovim inside our outside wsl (like the first post of this issue) when you see the hangs due to LSP?

I tried to repeat it with `neovide --wsl` but could not repeat it, I will try without wsl next, and a windows version of Biome.

Most likely this bug is in Neovim, it should make sure that all spawned processare are shutdown before exiting. There were similar problems with the clipboard servers not shutting down, but that is fixed.

**Comment by DeadRobotDev** [comment]
Sorry, I've been running it outside WSL when I see hangs due to LSP.

**Comment by fredizzimo** [comment]
I tried to repeat it on Windows with biome, but Neovide exits properly. Maybe it needs a bigger project, or some specific LSP configuration.

Do you have a git repository with a project that causes it to hang?

**Comment by fredizzimo** [comment]
Or maybe the way biome is installed matters, mine is installed through scoop.

**Comment by DeadRobotDev** [comment]
Example Project: https://github.com/DeadRobotDev/template-react-ts

I have Biome installed via Mason. I have no specific LSP config for Biome.

Even stripping my LSP config down to the bare minimum, I get the same issue.
```
return {
	'neovim/nvim-lspconfig',
	dependencies = {
		'williamboman/mason.nvim',
		'williamboman/mason-lspconfig.nvim',
	},
	event = {
		'BufReadPre',
		'BufNewFile',
	},
	opts = {
		biome = {},
	},
	config = function(_, opts)
		local lspconfig = require('lspconfig')

		require('mason').setup({
			ui = {
				border = 'rounded',
			},
		})

		require('mason-lspconfig').setup({
			automatic_installation = {
				exclude = {
				},
			},
		})

		for server, config in pairs(opts) do
			lspconfig[server].setup(config)
		end
	end,
}
```

**Comment by fredizzimo** [comment]
Thanks, I can reproduce it now with the biome installed through mason. The one installed through scoop does not have the issue though.

**Comment by fredizzimo** [comment]
I'm not sure why, and how to fix it, but it looks like Neovime passes the io pipes that are supposed to be for the communication between Neovide and Neovim forward to Node and Biome

![image](https://github.com/user-attachments/assets/49a4a182-35d8-4ed8-a49d-d3f7ecf64bc9)

The uv pipes are the ones that it's supposed to use for the communication between Neovim and Biome, but I don't know why the rust pipes are there.

**Comment by fredizzimo** [comment]
It's likely that a similar fix as this needs to be done https://github.com/neovim/neovim/pull/28157

`DuplicateHandle` has a parameter to prevent child processes from inheriting it. https://learn.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-duplicatehandle

**Comment by fredizzimo** [comment]
The problem might be more complex, libuv specifies that the handles are inherited here for some reason
https://github.com/libuv/libuv/blob/27134547ff7754107e2c490e0c2f0ed18c05dae2/src/win/process.c#L1057-L1070

And there does not seem to be a way to prevent that. 

We might be able to prevent it by changing the security descriptor, but that's all done deep inside the rust standard library and it's set to NULL, which means a default one. I'm not sure if we can change it.
https://learn.microsoft.com/en-us/windows/win32/sysinfo/handle-inheritance

**Comment by fredizzimo** [comment]
It might be possible to close the handles on the Neovide side, once we have detected that Neovim has exited, and there's nothing left to read.

**Comment by fredizzimo** [comment]
There's a pull request open on libuv that would fix the issue, if Neovim also specifies the correct flags to not inherit the handles, https://github.com/libuv/libuv/pull/3856. But unfortunately it has not been merged, despite being open for two years. It also creates other issues like https://github.com/libuv/libuv/issues/1490

**Comment by fredizzimo** [comment]
I created a quick workaround for this here:
* https://github.com/neovide/neovide/pull/2765

**Comment by vtjnash** [comment]
> The problem might be more complex, libuv specifies that the handles are inherited here for some reason

There might be a mild misunderstanding of the API here. That flag specifies that the CreateProcess is permitted to pass any handles (e.g. stdin/stdout/stderr). As such, that flag is basically useless.

> It's likely that a similar fix as this needs to be done https://github.com/neovim/neovim/pull/28157

After calling the MSVC `dup` function, you can call 
`SetHandleInformation(uv_get_osfhandle(handle), HANDLE_FLAG_INHERIT, 0);`
to do the equivalent of the race-y version of unix. Otherwise, you can implement a non-race-y `dup` directly with a combination of `uv_get_osfhandle`, `DuplicateHandle`, and `uv_open_osfhandle`. For a sample implementation, see `uv_convert_fd_to_handle` in libuv v2.

**Comment by fredizzimo** [comment]
@vtjnash thank you. 

I haven't had time to test this yet. But I will try to implement this in Neovim at some point and send a PR if it works.
